abdiff/core/collate_ab_transforms.py

```diff
@@ -0,0 +1,220 @@
+    TIMDEX records and compiling the dictionaries into a
+    pyarrow.RecordBatch. The size of a batch is set by the
+    'READ_BATCH_SIZE' global variable.
+
+    The function returns a generator, yielding batches of the dictionaries
+    from yield_records(). The returned generator can be passed to
+    abdiff.core.utils.write_to_dataset() to perform batch writes to
+    a Parquet dataset.
+    """
+    for transformed_files_list in transformed_files:
+        for transformed_file in transformed_files_list:
+            record_iter = yield_records(
+                transformed_file=Path(run_directory) / transformed_file
+            )
+            for record_batch in itertools.batched(record_iter, READ_BATCH_SIZE):
+                yield pa.RecordBatch.from_pylist(list(record_batch))
+
+
+def get_joined_batches_iter(dataset_directory: str) -> Generator:
+    """Yield pyarrow.RecordBatch objects of joined TIMDEX A/B records.
+
+    This function uses DuckDB to query the Parquet dataset of transformed
+    TIMDEX record dictionaries. It's worth noting that this Parquet dataset
+    is stored in a tempfile.TemporaryDirectory that gets deleted after the
+    function exits.
+
+    The following steps are performed in sequential order:
+        1. A list of DISTINCT transformed filenames (without the file extension)
+           is retrieved.
+
+        Steps 2-4 are performed for each transformed file.
+
+        2. Using common table expressions (CTEs) execute a query that performs
+           the following:
+           - Create a CTE called 'transformed_file' that contains all the records
+             where the partition transformed_file_name=the name of the transformed file.
+           - Create a CTE called 'a' that contains all the records from CTE
+             'transformed_file' where the column version='a'.
+           - Create a CTE called 'b' that contains all the records from CTE
+             'transformed_file' where the column version='b'.
+           - Join the 'a' and 'b' CTEs using the column timdex_record_id.
+
+        3. Read results from the query one batch at a time using fetch_record_batch().
+           Note: This will return the results as a pyarrow.RecordBatchReader.
+
+        4. Yield batches of the joined records until the pyarrow.RecordBatchReader
+           is empty.
+
+    The returned generator can be passed to abdiff.core.utils.write_to_dataset()
+    to perform batch writes of joined TIMDEX A/B records to a Parquet dataset
+
+    Args:
+        dataset_directory: The root directory of the Parquet dataset of TIMDEX records
+            (i.e., the tempfile.TemporaryDirectory).
+    """
+    with duckdb.connect() as con:
+        transformed_file_names = con.execute(
+            f"""
+            SELECT DISTINCT transformed_file_name FROM
+            read_parquet('{dataset_directory}/**/*.parquet', hive_partitioning=true)
+            """
+        ).fetchall()[0]
```

Comment:
As discussed in a quick huddle, this `[0]` index call is problematic if there are **multiple** transform files as part of the run.

Had discussed a couple of options are:
1. list comprehension to pull out the transform file name from each row returned
2. install `pandas` and use `.fetchdf()[COLUMN_NAME]` to get a list of values for that column in the SQL query

Thankfully, quite minor!  With that change, confirmed that the `for transformed_file in transformed_file_names` block later in this method successfully yields batches from each transform file, resulting in an expected number of rows in the final dataset.